### PR TITLE
Add Burn Contribution Method

### PR DESCRIPTION
**Pull request by amrro**
## Linear Ticket

[PRO-299 - Add delete contribution method to Contribution Resolver](https://linear.app/govrn/issue/PRO-299/add-delete-contribution-method-to-contribution-resolver)

## Description
Support contribution deletion regard the contribution status: 
- Add method that allows a user to delete a `staged` contribution
- And if the contribution is `minted`, it burns the contribution

contains a db migration.

### CODE REVIEW

File: libs/govrn-contract-client/src/lib/govrn-contract-client.ts [hunk 1]
```diff
@@ -78,4 +79,10 @@ export class GovrnContract {
   public async attestations(args: AttestationArgs) {
     return await this.govrn.attestations(args.tokenId, args.address);
   }
+
+  public async burnContribution(args: ContributionsArgs) {
+    return await this.govrn.burnContribution(args.tokenId, {
+      gasLimit: 2100000,
```
Comment (amrro) [inline libs/govrn-contract-client/src/lib/govrn-contract-client.ts:85]: @alexkeating is there any workaround for that?

File: apps/protocol-frontend/src/contexts/UserContext.tsx [hunk 2]
```diff
@@ -312,6 +313,42 @@ export const UserContextProvider: React.FC<UserContextProps> = ({
...
+            name: networks[chainId].name,
+          },
+          signer,
+          id
+        );
+        await getUserContributions();
+
+        toast({
+          title: 'Contribution Successfully deleted',
+          description: 'Your Contribution has been minted.',
```
Comment (alexkeating) [inline apps/protocol-frontend/src/contexts/UserContext.tsx:332]: Description does not match

File: libs/govrn-contract-client/src/lib/govrn-contract-client.ts (hunk 1 above)
Comment (alexkeating) [reply libs/govrn-contract-client/src/lib/govrn-contract-client.ts:85]: What happens if this is not specified?

File: libs/protocol-client/src/lib/client/contribution.ts [hunk 3]
```diff
@@ -29,6 +29,29 @@ export class Contribution extends BaseClient {
     return contributions.createContribution;
   }
 
+  public async burn(
```
Comment (alexkeating) [inline libs/protocol-client/src/lib/client/contribution.ts:32]: Probably better to call this delete even though it burns on chain

File: libs/protocol-client/src/lib/client/contribution.ts [hunk 4]
```diff
@@ -29,6 +29,29 @@ export class Contribution extends BaseClient {
...
+  ) {
+    const contract = new GovrnContract(networkConfig, provider);
+
+    const contribution = await this.get(id);
+    if (!contribution) {
+      throw Error(`Contribution doesn't exist: ${id}`);
+    }
+
+    // TODO: Can we avoid hardcoding the event name
+    if (contribution?.status.name === 'minted' && contribution.on_chain_id) {
```
Comment (alexkeating) [inline libs/protocol-client/src/lib/client/contribution.ts:45]: Good call, maybe we should create a constants library

**Review by alexkeating — COMMENTED** [review]
This looks great! A couple small comments but otherwise good

File: apps/protocol-api/src/main.ts [hunk 5]
```diff
@@ -80,6 +81,7 @@ const permissions = shield(
       createUserContribution: and(OwnsData, isAuthenticated),
       createUserCustom: or(hasToken, and(OwnsData, isAuthenticated)),
       createUserOnChainAttestation: isAuthenticated,
+      deleteContribution: or(OwnsData, isAuthenticated),
```
Comment (alexkeating) [inline apps/protocol-api/src/main.ts:84]: We are going to need a custom resolver that checks the user address owns the contribution they are deleting. Right now any logged in user can delete any contribution

File: libs/govrn-contract-client/src/lib/govrn-contract-client.ts (hunk 1 above)
Comment (amrro) [reply libs/govrn-contract-client/src/lib/govrn-contract-client.ts:85]: [`UNPREDICTABLE_GAS_LIMIT`](https://docs.ethers.io/v5/troubleshooting/errors/#help-UNPREDICTABLE_GAS_LIMIT) error.

Comment (alexkeating) [reply libs/govrn-contract-client/src/lib/govrn-contract-client.ts:85]: RInkeby

File: apps/protocol-api/src/prisma/resolvers/Contribution.ts [hunk 6]
```diff
@@ -423,4 +440,32 @@ export class ContributionCustomResolver {
...
+        id: { equals: contributionId },
+      },
+    });
+
+    if (query.length === 0) {
+      throw Error("Contribution doesn't exist.");
+    }
+
+    const contribution = query[0];
+    if (address === contribution.user.address) {
```
Comment (alexkeating) [inline apps/protocol-api/src/prisma/resolvers/Contribution.ts:463]: !==

**Pull request merged**